python/ray/tune/trial_runner.py

```diff
@@ -414,12 +414,28 @@ def _process_events(self):
             if trial.is_restoring:
                 with warn_if_slow("process_trial_restore"):
                     self._process_trial_restore(trial)
+            elif trial.is_saving:
+                with warn_if_slow("process_trial_save") as profile:
+                    self._process_trial_save(trial)
+                if profile.too_slow and trial.sync_on_checkpoint:
+                    # TODO(ujvl): Suggest using DurableTrainable once
+                    #  API has converged.
+                    logger.warning(
+                        "Consider turning off forced head-worker trial "
+                        "checkpoint syncs by setting sync_on_checkpoint=False"
+                        ". Note that this may result in faulty trial "
+                        "restoration if a failure occurs while the checkpoint "
+                        "is being synced from the worker to the head node.")
             else:
                 with warn_if_slow("process_trial"):
                     self._process_trial(trial)
 
     def _process_trial(self, trial):
-        """Processes a trial result."""
+        """Processes a trial result.
+
+        Args:
+            trial (Trial): Trial with a result ready to be processed.
```

Comment:
Can you actually document the below logic flow in this docstring?

This will make it much easier for others to develop/follow. Specifically, it should answer questions like

1. how does the async saving work?
2. why is there a decision cache?

Reply:
added to docstring and comments, lmk if more detail required